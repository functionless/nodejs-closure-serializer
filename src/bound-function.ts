import { globalSingleton } from "./global-singleton";

export interface BoundFunction {
  "[[TargetFunction]]": Function;
  "[[BoundThis]]": any;
  "[[BoundArgs]]": any[];
}

const boundFunctions = globalSingleton(
  Symbol.for("@functionless/bound-functions"),
  () => new WeakMap<Function, BoundFunction>()
);

export function getBoundFunction(func: Function): BoundFunction | undefined {
  return boundFunctions.get(func);
}

const bind = Function.prototype.bind;

Function.prototype.bind = function (boundThis: any, ...boundArgs: any[]) {
  const bound = bind.call(this, boundThis, ...boundArgs);
  boundFunctions.set(bound, {
    "[[TargetFunction]]": this,
    "[[BoundThis]]": boundThis,
    "[[BoundArgs]]": boundArgs,
  });
  return bound;
};
