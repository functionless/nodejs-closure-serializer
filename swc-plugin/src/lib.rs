use std::collections::{HashMap};
use swc_common::{chain, Mark};
use swc_common::util::take::Take;
use swc_ecma_visit::Fold;
use swc_plugin::{ast::*, plugin_transform, TransformPluginProgramMetadata, utils::*};

#[plugin_transform]
pub fn wrap_closures(mut program: Program, _metadata: TransformPluginProgramMetadata) -> Program {
    program.visit_mut_with(&mut ClosureSerializer {
        stack: LexicalScope::new()
    });

    program
}

pub fn wrap(top_level_mark: Mark) -> impl Fold + VisitMut {
    as_folder(ClosureSerializer {
        stack: LexicalScope::new()
    })
}

pub struct ClosureSerializer {
    /**
     * The [lexical scope](LexicalScope) of the program at the current point of the AST.
     */
    stack: LexicalScope
}

impl ClosureSerializer {
    /**
     * Generic function that will walk all statements in a block and hoist
     * all function declarations and any var declarations that can be hoisted.
     * 
     * Stores the names produced by a [stmt](Stmt):
     * 1. function declarations
     * ```ts
     * function foo() {}
     * ```
     * 2. var declarations that have no initializer
     * ```ts
     * var foo;
     * ```
     */
    fn bind_hoisted_stmts_in_block<T>(&mut self, block: &mut Vec<T>)
    where
        T: StmtLike + VisitMutWith<Self>,
    {
        block.into_iter().for_each(|stmt| {
            // hoist all of the function and var declarations in the module into scope
            match stmt.as_stmt() {
                Some(stmt) => {
                    match stmt {
                        Stmt::Decl(Decl::Var(var)) => {
                            if var.kind == VarDeclKind::Var {
                                for decl in var.decls.iter() {
                                    if decl.init.is_none() {
                                        // var declarations with no initialized are always hoisted
                                        self.stack.bind_pat(&decl.name);
                                    }
                                }
                            }
                        }
                        Stmt::Decl(Decl::Fn(func)) => {
                            self.stack.bind_ident(&func.ident);
                        }
                        _ => {}
                    }
                    // self.scope.hoist_stmt(&stmt);
                },
                _ => {}
            }
        });
    }
}

impl VisitMut for ClosureSerializer {
    // Implement necessary visit_mut_* methods for actual custom transform.
    // A comprehensive list of possible visitor methods can be found here:
    // https://rustdoc.swc.rs/swc_ecma_visit/trait.VisitMut.html

    fn visit_mut_module_items(&mut self, items: &mut Vec<ModuleItem>) {
        self.bind_hoisted_stmts_in_block(items);
        items.iter_mut().for_each(|stmt| stmt.visit_mut_with(self));
    }

    fn visit_mut_block_stmt(&mut self, block: &mut BlockStmt) {
        // we are entering a block, so push a frame onto the stack
        self.stack.push();

        self.bind_hoisted_stmts_in_block(&mut block.stmts);

        // now that all hoisted variables are in scope, walk each of the children
        block.visit_mut_children_with(self);

        // finally, pop the stack frame
        self.stack.pop();
    }

    fn visit_mut_var_decl(&mut self, var: &mut VarDecl) {
        for decl in var.decls.iter_mut() {
            match decl.init.as_deref_mut() {
                Some(init) => {
                    // var x = v;
                    // let x = b;
                    // const x = v;

                    // bind the names to the current lexical scope
                    self.stack.bind_pat(&decl.name);

                    // then visit the initializer with the updated lexical scope
                    init.visit_mut_with(self);
                }
                None if var.kind == VarDeclKind::Var => {
                    // hoisted var - we should ignore as it has already been hoisted at the beginning of the block
                    // var x;
                }
                None => {
                    // let x;

                    // bind the names to the current lexical scope
                    self.stack.bind_pat(&decl.name);
                }
            }
        }
    }

    fn visit_mut_expr(&mut self, expr: &mut Expr) {
        match expr {
            Expr::Arrow(arrow) => {
                // push a new frame onto the stack for the contents of this function
                self.stack.push();

                arrow.params.iter_mut().for_each(|param| {
                    // bind this argument into lexical scope

                    self.stack.bind_pat(param);
                    match param {
                        Pat::Assign(assign) => {
                            // this is a parameter with a default value
                            // e.g (a, b = a)
                            // or  (a, b = () => [a, b])

                            // we must transform the initializer with the arguments to its left in scope
                            assign.right.as_mut().visit_mut_children_with(self);
                        }
                        _ => {}
                    }
                });


                if arrow.body.is_expr() {
                  let expr = arrow.body.as_mut_expr().unwrap();

                  expr.visit_mut_with(self);
                } else {
                  let block = arrow.body.as_mut_block_stmt().unwrap();

                  // hoist all of the function/var declarations into scope
                  self.bind_hoisted_stmts_in_block(&mut block.stmts);

                  // process each of the children
                  block.visit_mut_children_with(self);
                }

                // global.wrapClosure((...args) => { ..stmts })
                let call = CallExpr {
                    span: arrow.span,
                    callee: Callee::Expr(Box::new(
                        Expr::Member(MemberExpr {
                            obj: Box::new(Expr::Ident(private_ident!(arrow.span, "global"))),
                            prop: MemberProp::Ident(private_ident!(arrow.span, "wrapClosure")),
                            span: arrow.span
                        })
                    )),
                    args: vec!(ExprOrSpread {
                        expr: Box::new(Expr::Arrow(arrow.take())),
                        spread: None
                    }), // TODO: inject metadata about free variables
                    type_args: None
                };

                // replace the ArrowExpr with a call to wrapClosure, wrapping the ArrowExpr with metadata
                *expr = Expr::Call(call);

                self.stack.pop();
            },
            Expr::Fn(function) => {

            }
            _ => {}
        }
    }
}

/**
 * A mapping of [reference name](JsWord) to the [unique id](u32) of that reference.
 */
type Frame = HashMap<JsWord, u32>;

struct LexicalScope {
    /**
     * Counter for assigning unique identifiers.
     */
    count: u32,
    /**
     * Mapping of a [reference](Id) to its assigned unique id.
     */
    ids: HashMap<Id, u32>,
    /**
     * A list of [stack frames](Frame) for the program at the current point in the tree.
     */
    stack: Vec<Frame>
}

impl LexicalScope {
    pub fn new() -> Self {
        LexicalScope { 
            count: 0,
            ids: HashMap::new(),
            stack: vec!(Frame::new())
        }
    }

    /**
     * Walk backwards through the Scope chain to find the variable id.
     */
    fn lookup(&self, name: &JsWord) -> Option<u32> {
        for scope in self.stack.iter() {
            let val = scope.get(name);
            if val.is_some() {
                return val.cloned();
            }
        }
        Option::None
    }

    fn frame(&mut self) -> &mut Frame {
        self.stack.last_mut().unwrap()
    }

    /**
     * Push a Scope onto the Stack.
     */
    fn push(&mut self) -> &mut Frame {
        self.stack.push(HashMap::new());
        self.stack.last_mut().unwrap()
    }

    fn pop(&mut self) {
        if self.stack.pop().is_none() {
            panic!("stack underflow");
        }
    }

    /**
     * Binds the name of an [ident](Ident) to the current [lexical scope](LexicalScope).
     */
    fn bind_ident(&mut self, ident: &Ident) {
        let id = self.get_unique_id(&ident);
        self.frame().insert(ident.to_id().0, id);
    }

    /**
     * Get (or assign) a [unique id](u32) for an [identifier](Ident).
     * 
     * The ID will be used to uniquely identify a variable (regardless of name shadowing/collisions).
     */
    fn get_unique_id(&mut self, ident: &Ident) -> u32 {
        let id = ident.to_id();
        if !self.ids.contains_key(&id) {
            self.count += 1;
            self.ids.insert(id, self.count);
        }
        *self.ids.get(&ident.to_id()).unwrap()
    }

    /**
     * Binds the names produced by a [binding pattern](Pat) to the current [lexical scope](LexicalScope).
     * 
     * ```ts
     * // patterns:
     * a
     * {b}
     * {d: c}
     * [d];
     * ```
     */
    fn bind_pat(&mut self, pat: &Pat) {
        match pat {
            Pat::Ident(ident) => {
                self.bind_ident(&ident.id);
            },
            Pat::Object(o) => {
                for prop in o.props.iter() {
                    match prop {
                        ObjectPatProp::Assign(a) => {
                            self.bind_ident(&a.key);
                        }
                        ObjectPatProp::KeyValue(kv) => {
                            self.bind_pat(kv.value.as_ref());
                        }
                        _ => {}
                    }
                };
            },
            Pat::Array(a) => {
                for element in a.elems.iter() {
                    if element.is_some() {
                        self.bind_pat(element.as_ref().unwrap());
                    }
                }
            }
            _ => {}
        }
    }
}
