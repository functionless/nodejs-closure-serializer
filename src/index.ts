// ensure that the Function.prototype.bind function has been patched
import "./bound-function";

export { serializeFunction, SerializeFunctionProps } from "./serialize-closure";
