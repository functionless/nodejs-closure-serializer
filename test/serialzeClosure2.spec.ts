import fs from "fs";
import path from "path";
import * as uuid from "uuid";
import { serializeFunction, SerializeFunctionProps } from "../src";

test("maintain captured value", async () => {
  let i = 0;

  function foo() {
    i += 1;
  }

  function bar() {
    return i;
  }

  const handler = () => {
    foo();
    return bar();
  };

  await serializeFunction(handler);
});

test("capturing a reference to a function", () => {
  function foo() {
    return "hello";
  }

  return testCase({
    closure: () => foo(),
    expectResult: "hello",
  });
});

test("capturing a reference to a string", () => {
  const foo = "hello";

  return testCase({
    closure: () => foo,
    expectResult: foo,
  });
});

test("capturing a reference to an array", () => {
  const foo = ["hello"];

  return testCase({
    closure: () => foo,
    expectResult: foo,
  });
});

test("capturing a reference to an array containing a function", () => {
  function bar() {
    return "hello";
  }

  const foo = [bar];

  return testCase({
    closure: () => foo,
    expectResult: [expect.any(Function)],
  });
});

test("value captured multiple times is only emitted once", () => {
  function bar() {
    return "hello";
  }

  const b = bar; // even if the value is captured indirectly
  const foo = [bar, bar, b];

  return testCase({
    closure: () => foo,
    expectResult: [
      expect.any(Function),
      expect.any(Function),
      expect.any(Function),
    ],
  });
});

test("capturing a reference to a native bound function", () => {
  function foo() {
    return this.internal;
  }

  const f = foo.bind({
    internal: "value",
  });

  return testCase({
    closure: () => f(),
    expectResult: "value",
  });
});

test("capturing a reference to a native bound method", () => {
  class F {
    internal: string;
    foo(a, { b }, [c]) {
      return [this.internal, a, b, c];
    }
  }

  const f = new F();
  const func = f.foo.bind({
    internal: "value",
  });

  return testCase({
    closure: () => func("a", { b: "b" }, ["c"]),
    expectResult: ["value", "a", "b", "c"],
  });
});

test("capturing a reference to a native bound async method", () => {
  class F {
    internal: string;
    async foo() {
      return this.internal;
    }
  }

  const f = new F();
  const func = f.foo.bind({
    internal: "value",
  });

  return testCase({
    closure: () => func(),
    expectResult: "value",
  });
});

test("capturing a reference to a native bound generator method", () => {
  class F {
    i: number = 0;
    *foo() {
      while ((this.i += 1) <= 1) {
        yield "hello";
      }
    }
  }

  const f = new F();
  const func = f.foo.bind(f);

  return testCase({
    closure: () => func(),
    expectResult: (generator: any) => {
      const values: any[] = [];
      for (const i of generator) {
        values.push(i);
      }
      expect(values).toEqual(["hello"]);
    },
  });
});

test("arrow function nested within a function", () => {
  function foo() {
    return ((val) => `${val} ${this.internal}`)("hello");
  }

  return testCase({
    closure: foo.bind({ internal: "value" }),
    expectResult: "hello value",
  });
});

test("class method", () => {
  class Foo {
    constructor(readonly internal: string) {}
    foo() {
      return this.internal;
    }
  }

  const foo = new Foo("hello");

  return testCase({
    closure: () => foo.foo(),
    expectResult: "hello",
  });
});

test("super class method", () => {
  class Foo {
    constructor(readonly internal: string) {}
    foo() {
      return this.internal;
    }
  }
  class Bar extends Foo {}

  const foo = new Bar("hello");

  return testCase({
    closure: () => foo.foo(),
    expectResult: "hello",
  });
});

test("call method on class with prototype swapped", () => {
  class A {
    constructor(readonly internal: string) {}

    foo() {
      return `${this.internal} a`;
    }
  }
  class B {
    constructor(readonly internal: string) {}

    foo() {
      return `${this.internal} b`;
    }
  }

  class C extends A {}
  Object.setPrototypeOf(C, B);
  Object.setPrototypeOf(C.prototype, B.prototype);

  return testCase({
    closure: () => new C("value").foo(),
    expectResult: "value b",
  });
});

test("call static method on class with prototype swapped", () => {
  class A {
    static foo() {
      return "A";
    }
  }
  class B {
    static foo() {
      return "B";
    }
  }

  class C extends A {}
  Object.setPrototypeOf(C, B);
  Object.setPrototypeOf(C.prototype, B.prototype);

  return testCase({
    closure: () => C.foo(),
    expectResult: "B",
  });
});

test("class mix-in", () => {
  const mixin = (Base: new (internal: string) => any) =>
    class extends Base {
      constructor(internal: string) {
        super(internal);
      }
      foo() {
        return this.internal;
      }
    };

  const A = mixin(
    class {
      constructor(readonly internal: string) {}
    }
  );

  const a = new A("value");
  return testCase({
    closure: () => a.foo(),
    expectResult: "value",
  });
});

test("call method on traditional function prototype class", () => {
  function Animal(noise: string) {
    this.noise = noise;
  }
  Animal.prototype.speak = function () {
    return this.noise;
  };
  const animal = new Animal("bork");

  function Dog(noise: string) {
    Animal.call(this, `bark ${noise}`);
  }
  Object.setPrototypeOf(Dog.prototype, Animal.prototype);

  return testCase({
    closure: () => [new Dog("woof").speak(), animal.speak()],
    expectResult: ["bark woof", "bork"],
  });
});

test("call the exports.handler if isFactoryFunction", () => {
  function expensiveTask() {
    return "world";
  }

  return testCase({
    closure: () => {
      const expensive = expensiveTask();

      return (handleInput: string) => {
        return `${handleInput} ${expensive}`;
      };
    },
    args: ["hello"],
    expectResult: "hello world",
    isFactoryFunction: true,
  });
});

test("all binding patterns should be considered when detecting free variables", () => {
  const free = "free";

  class FreeClass {}

  return testCase({
    closure: (
      // argument identifier
      argument: string,
      // object binding pattern in argument
      { objArgument }: any,
      // array binding pattern in argument
      [arrayArgument]: any,
      // nested array binding pattern in argument
      { a: [arrayInObjectArgument] },
      // nested object binding pattern in argument
      { b: { objectBindingInObjectBindingArgument } }
    ) => {
      function notHoisted() {
        return "notHoisted";
      }

      const arrowFunction = () => "arrowFunction";

      const functionExpression = function () {
        return "functionExpression";
      };

      class ClassDeclaration extends FreeClass {
        //                              ^ free

        constructor(readonly text: string = "ClassDeclaration") {
          super();
        }

        get() {
          return "ClassDeclaration";
        }
      }

      const ClassExpression = class extends FreeClass {
        //                                    ^ free
        get() {
          return "ClassExpression";
        }
      };

      const NamedClassExpression = class NamedClassExpression extends FreeClass {
        //                                                              ^ free
        get() {
          return "NamedClassExpression";
        }
      };

      function SuperClass(text: string) {
        this.text = text;
      }
      SuperClass.prototype.get = function () {
        return this.text;
      };
      function TraditionalClass() {
        SuperClass.call(this, "TraditionalClass");
      }
      Object.setPrototypeOf(TraditionalClass.prototype, SuperClass.prototype);

      // single variable declaration with ts.Identifier
      const id = "id";

      // variable declaration list
      const a = "a",
        b = "b";

      // object binding pattern
      const {
        objPattern1,
        a: [arrayPatternInObject],
      } = {
        objPattern1: "objPattern1",
        a: ["arrayPatternInObject"],
      };
      // object binding patterns in a variable declaration list
      const { objPatternList1 } = {
          objPatternList1: "objPatternList1",
        },
        { objPatternList2 } = {
          objPatternList2: "objPatternList2",
        };

      // array binding pattern
      const [arrayPattern1, { objPatternInArray }] = [
        "arrayPattern1",
        { objPatternInArray: "objPatternInArray" },
      ];

      // array binding pattern list
      const [arrayPatternList1] = ["arrayPatternList1"],
        [arrayPatternList2] = ["arrayPatternList2"];

      return [
        argument,
        objArgument,
        arrayArgument,
        arrayInObjectArgument,
        objectBindingInObjectBindingArgument,
        free,
        id,
        a,
        b,
        objPattern1,
        arrayPatternInObject,
        objPatternList1,
        objPatternList2,
        arrayPattern1,
        objPatternInArray,
        arrayPatternList1,
        arrayPatternList2,
        notHoisted(),
        hoisted(),
        arrowFunction(),
        functionExpression(),
        new ClassDeclaration().get(),
        new ClassExpression().get(),
        new NamedClassExpression().get(),
        new TraditionalClass().get(),
      ];

      function hoisted() {
        return "hoisted";
      }
    },
    args: [
      "argument",
      { objArgument: "objArgument" },
      ["arrayArgument"],
      { a: ["arrayInObjectArgument"] },
      {
        b: {
          objectBindingInObjectBindingArgument:
            "objectBindingInObjectBindingArgument",
        },
      },
    ],
    expectResult: [
      "argument",
      "objArgument",
      "arrayArgument",
      "arrayInObjectArgument",
      "objectBindingInObjectBindingArgument",
      "free",
      "id",
      "a",
      "b",
      "objPattern1",
      "arrayPatternInObject",
      "objPatternList1",
      "objPatternList2",
      "arrayPattern1",
      "objPatternInArray",
      "arrayPatternList1",
      "arrayPatternList2",
      "notHoisted",
      "hoisted",
      "arrowFunction",
      "functionExpression",
      "ClassDeclaration",
      "ClassExpression",
      "NamedClassExpression",
      "TraditionalClass",
    ],
  });
});

test("should not capture global values as free variables", () => {
  return testCase({
    closure: () => {
      // Object.getOwnPropertyNames(global).sort().join(';\n')

      // $0;
      // $1;
      // $2;
      // $3;
      // $4;
      // $_;
      AbortController;
      AbortSignal;
      // AggregateError;
      Array;
      ArrayBuffer;
      Atomics;
      BigInt;
      BigInt64Array;
      BigUint64Array;
      Boolean;
      Buffer;
      DataView;
      Date;
      Error;
      EvalError;
      Event;
      EventTarget;
      // FinalizationRegistry;
      Float32Array;
      Float64Array;
      Function;
      Infinity;
      Int16Array;
      Int32Array;
      Int8Array;
      Intl;
      JSON;
      Map;
      Math;
      NaN;
      Number;
      Object;
      Promise;
      Proxy;
      RangeError;
      ReferenceError;
      Reflect;
      RegExp;
      Set;
      SharedArrayBuffer;
      String;
      Symbol;
      SyntaxError;
      TextDecoder;
      TextEncoder;
      TypeError;
      URIError;
      URL;
      URLSearchParams;
      Uint16Array;
      Uint32Array;
      Uint8Array;
      Uint8ClampedArray;
      WeakMap;
      // WeakRef;
      WeakSet;
      WebAssembly;
      // afterAll;
      // afterEach;
      atob;
      // beforeAll;
      // beforeEach;
      btoa;
      // clear;
      clearImmediate;
      clearInterval;
      clearTimeout;
      console;
      // copy;
      // debug;
      decodeURI;
      decodeURIComponent;
      describe;
      // dir;
      // dirxml;
      encodeURI;
      encodeURIComponent;
      escape;
      eval;
      // expect;
      // fdescribe;
      // fit;
      global;
      globalThis;
      // inspect;
      isFinite;
      isNaN;
      // it;
      // jest-symbol-do-not-touch;
      // keys;
      // monitor;
      parseFloat;
      parseInt;
      performance;
      process;
      // profile;
      // profileEnd;
      // queryObjects;
      queueMicrotask;
      // require; // can't test this, stupid jest overrides it and we end up traversing all of JEST
      setImmediate;
      setInterval;
      setTimeout;
      // table;
      // test;
      // ts - jest;
      // undebug;
      undefined;
      unescape;
      // unmonitor;
      // values;
      // xdescribe;
      // xit;
      // xtest;
    },
    expectResult: undefined,
  });
});

async function testCase<
  F extends (...args: any[]) => any,
  T,
  IsFactoryFunction extends boolean = false
>(testCase: {
  preSerializeValue?: SerializeFunctionProps["preSerializeValue"];
  closure: F;
  isFactoryFunction?: IsFactoryFunction;
  args?: any;
  expectResult: T | ((t: T) => void);
}) {
  const { closure, args, expectResult, isFactoryFunction, preSerializeValue } =
    testCase;
  const serialized = await serializeFunction(closure, {
    isFactoryFunction,
    preSerializeValue,
  });
  expect(serialized).toMatchSnapshot();
  const fileName = path.join(__dirname, `${uuid.v4()}.js`);
  try {
    fs.writeFileSync(fileName, serialized);
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    let js = require(fileName).handler;
    let actualResult: any = js(...(args ?? []));
    if (typeof actualResult?.then === "function") {
      actualResult = await actualResult;
    }
    if (typeof expectResult === "function") {
      (<any>expectResult)(actualResult);
    } else {
      expect(actualResult).toEqual(expectResult);
    }
  } finally {
    fs.rmSync(fileName);
  }
}
