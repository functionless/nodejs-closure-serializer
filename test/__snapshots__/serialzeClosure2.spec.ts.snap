// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all binding patterns should be considered when detecting free variables 1`] = `
"const v1 = (() => class FreeClass {
})();
const v2 = ((FreeClass, free) => (
// argument identifier
argument, 
// object binding pattern in argument
{ objArgument }, 
// array binding pattern in argument
[arrayArgument], 
// nested array binding pattern in argument
{ a: [arrayInObjectArgument] }, 
// nested object binding pattern in argument
{ b: { objectBindingInObjectBindingArgument } }) => {
    function notHoisted() {
        return \\"notHoisted\\";
    }
    const arrowFunction = () => \\"arrowFunction\\";
    const functionExpression = function () {
        return \\"functionExpression\\";
    };
    class ClassDeclaration extends FreeClass {
        //                              ^ free
        constructor(text = \\"ClassDeclaration\\") {
            super();
            this.text = text;
        }
        get() {
            return \\"ClassDeclaration\\";
        }
    }
    const ClassExpression = class extends FreeClass {
        //                                    ^ free
        get() {
            return \\"ClassExpression\\";
        }
    };
    const NamedClassExpression = class NamedClassExpression extends FreeClass {
        //                                                              ^ free
        get() {
            return \\"NamedClassExpression\\";
        }
    };
    function SuperClass(text) {
        this.text = text;
    }
    SuperClass.prototype.get = function () {
        return this.text;
    };
    function TraditionalClass() {
        SuperClass.call(this, \\"TraditionalClass\\");
    }
    Object.setPrototypeOf(TraditionalClass.prototype, SuperClass.prototype);
    // single variable declaration with ts.Identifier
    const id = \\"id\\";
    // variable declaration list
    const a = \\"a\\", b = \\"b\\";
    // object binding pattern
    const { objPattern1, a: [arrayPatternInObject], } = {
        objPattern1: \\"objPattern1\\",
        a: [\\"arrayPatternInObject\\"],
    };
    // object binding patterns in a variable declaration list
    const { objPatternList1 } = {
        objPatternList1: \\"objPatternList1\\",
    }, { objPatternList2 } = {
        objPatternList2: \\"objPatternList2\\",
    };
    // array binding pattern
    const [arrayPattern1, { objPatternInArray }] = [
        \\"arrayPattern1\\",
        { objPatternInArray: \\"objPatternInArray\\" },
    ];
    // array binding pattern list
    const [arrayPatternList1] = [\\"arrayPatternList1\\"], [arrayPatternList2] = [\\"arrayPatternList2\\"];
    return [
        argument,
        objArgument,
        arrayArgument,
        arrayInObjectArgument,
        objectBindingInObjectBindingArgument,
        free,
        id,
        a,
        b,
        objPattern1,
        arrayPatternInObject,
        objPatternList1,
        objPatternList2,
        arrayPattern1,
        objPatternInArray,
        arrayPatternList1,
        arrayPatternList2,
        notHoisted(),
        hoisted(),
        arrowFunction(),
        functionExpression(),
        new ClassDeclaration().get(),
        new ClassExpression().get(),
        new NamedClassExpression().get(),
        new TraditionalClass().get(),
    ];
    function hoisted() {
        return \\"hoisted\\";
    }
})(v1, \\"free\\");
exports.handler = v2"
`;

exports[`arrow function nested within a function 1`] = `
"var v2 = { internal: \\"value\\" };
Object.setPrototypeOf(v2, Object.prototype);
const v1 = (_self => function foo() {
    return ((val) => \`\${val} \${this.internal}\`)(\\"hello\\");
}.bind(_self))(v2);
exports.handler = v1"
`;

exports[`call method on class with prototype swapped 1`] = `
"const v2 = (() => class B {
    constructor(internal) {
        this.internal = internal;
    }
    foo() {
        return \`\${this.internal} b\`;
    }
})();
const v1 = (A => class C extends A {
})(v2);
const v3 = (C => () => new C(\\"value\\").foo())(v1);
exports.handler = v3"
`;

exports[`call method on traditional function prototype class 1`] = `
"const v5 = (() => function () {
    return this.noise;
})();
var v3 = { speak: v5 };
Object.setPrototypeOf(v3, Object.prototype);
var v2 = { noise: \\"bork\\" };
Object.setPrototypeOf(v2, v3);
const v6 = (() => function Animal(noise) {
    this.noise = noise;
})();
const v1 = (Animal => function Dog(noise) {
    Animal.call(this, \`bark \${noise}\`);
})(v6);
const v7 = ((Dog, animal) => () => [new Dog(\\"woof\\").speak(), animal.speak()])(v1, v2);
exports.handler = v7"
`;

exports[`call static method on class with prototype swapped 1`] = `
"const v2 = (() => class B {
    static foo() {
        return \\"B\\";
    }
})();
const v1 = (A => class C extends A {
})(v2);
const v3 = (C => () => C.foo())(v1);
exports.handler = v3"
`;

exports[`call the exports.handler if isFactoryFunction 1`] = `
"const v1 = (() => function expensiveTask() {
    return \\"world\\";
})();
const v2 = (expensiveTask => () => {
    const expensive = expensiveTask();
    return (handleInput) => {
        return \`\${handleInput} \${expensive}\`;
    };
})(v1);
exports.handler = v2()"
`;

exports[`capturing a reference to a function 1`] = `
"const v1 = (() => function foo() {
    return \\"hello\\";
})();
const v2 = (foo => () => foo())(v1);
exports.handler = v2"
`;

exports[`capturing a reference to a native bound async method 1`] = `
"const v2 = (() => function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try {
            step(generator.next(value));
        }
        catch (e) {
            reject(e);
        } }
        function rejected(value) { try {
            step(generator[\\"throw\\"](value));
        }
        catch (e) {
            reject(e);
        } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
})();
var v4 = { internal: \\"value\\" };
Object.setPrototypeOf(v4, Object.prototype);
const v3 = ((_self, __awaiter) => function foo() {
    return __awaiter(this, void 0, void 0, function* () {
        return this.internal;
    });
}.bind(_self))(v4, v2);
const v6 = (func => () => func())(v3);
exports.handler = v6"
`;

exports[`capturing a reference to a native bound function 1`] = `
"var v3 = { internal: \\"value\\" };
Object.setPrototypeOf(v3, Object.prototype);
const v2 = (_self => function foo() {
    return this.internal;
}.bind(_self))(v3);
const v5 = (f => () => f())(v2);
exports.handler = v5"
`;

exports[`capturing a reference to a native bound generator method 1`] = `
"var v4 = {};
Object.setPrototypeOf(v4, Object.prototype);
var v3 = { i: 0 };
Object.setPrototypeOf(v3, v4);
const v2 = (_self => function* foo() {
    while ((this.i += 1) <= 1) {
        yield \\"hello\\";
    }
}.bind(_self))(v3);
const v6 = (func => () => func())(v2);
exports.handler = v6"
`;

exports[`capturing a reference to a native bound method 1`] = `
"var v3 = { internal: \\"value\\" };
Object.setPrototypeOf(v3, Object.prototype);
const v2 = (_self => function foo(a, { b }, [c]) {
    return [this.internal, a, b, c];
}.bind(_self))(v3);
const v5 = (func => () => func(\\"a\\", { b: \\"b\\" }, [\\"c\\"]))(v2);
exports.handler = v5"
`;

exports[`capturing a reference to a string 1`] = `
"const v1 = (foo => () => foo)(\\"hello\\");
exports.handler = v1"
`;

exports[`capturing a reference to an array 1`] = `
"const v2 = (foo => () => foo)([\\"hello\\"]);
exports.handler = v2"
`;

exports[`capturing a reference to an array containing a function 1`] = `
"const v2 = (() => function bar() {
    return \\"hello\\";
})();
const v3 = (foo => () => foo)([v2]);
exports.handler = v3"
`;

exports[`class method 1`] = `
"var v2 = {};
Object.setPrototypeOf(v2, Object.prototype);
var v1 = { internal: \\"hello\\" };
Object.setPrototypeOf(v1, v2);
const v4 = (foo => () => foo.foo())(v1);
exports.handler = v4"
`;

exports[`class mix-in 1`] = `
"var v3 = {};
Object.setPrototypeOf(v3, Object.prototype);
var v2 = {};
Object.setPrototypeOf(v2, v3);
var v1 = { internal: \\"value\\" };
Object.setPrototypeOf(v1, v2);
const v5 = (a => () => a.foo())(v1);
exports.handler = v5"
`;

exports[`should not capture global values as free variables 1`] = `
"const v1 = (() => () => {
    // Object.getOwnPropertyNames(global).sort().join(';\\\\n')
    // $0;
    // $1;
    // $2;
    // $3;
    // $4;
    // $_;
    AbortController;
    AbortSignal;
    // AggregateError;
    Array;
    ArrayBuffer;
    Atomics;
    BigInt;
    BigInt64Array;
    BigUint64Array;
    Boolean;
    Buffer;
    DataView;
    Date;
    Error;
    EvalError;
    Event;
    EventTarget;
    // FinalizationRegistry;
    Float32Array;
    Float64Array;
    Function;
    Infinity;
    Int16Array;
    Int32Array;
    Int8Array;
    Intl;
    JSON;
    Map;
    Math;
    NaN;
    Number;
    Object;
    Promise;
    Proxy;
    RangeError;
    ReferenceError;
    Reflect;
    RegExp;
    Set;
    SharedArrayBuffer;
    String;
    Symbol;
    SyntaxError;
    TextDecoder;
    TextEncoder;
    TypeError;
    URIError;
    URL;
    URLSearchParams;
    Uint16Array;
    Uint32Array;
    Uint8Array;
    Uint8ClampedArray;
    WeakMap;
    // WeakRef;
    WeakSet;
    WebAssembly;
    // afterAll;
    // afterEach;
    atob;
    // beforeAll;
    // beforeEach;
    btoa;
    // clear;
    clearImmediate;
    clearInterval;
    clearTimeout;
    console;
    // copy;
    // debug;
    decodeURI;
    decodeURIComponent;
    describe;
    // dir;
    // dirxml;
    encodeURI;
    encodeURIComponent;
    escape;
    eval;
    // expect;
    // fdescribe;
    // fit;
    global;
    globalThis;
    // inspect;
    isFinite;
    isNaN;
    // it;
    // jest-symbol-do-not-touch;
    // keys;
    // monitor;
    parseFloat;
    parseInt;
    performance;
    process;
    // profile;
    // profileEnd;
    // queryObjects;
    queueMicrotask;
    // require; // can't test this, stupid jest overrides it and we end up traversing all of JEST
    setImmediate;
    setInterval;
    setTimeout;
    // table;
    // test;
    // ts - jest;
    // undebug;
    undefined;
    unescape;
    // unmonitor;
    // values;
    // xdescribe;
    // xit;
    // xtest;
})();
exports.handler = v1"
`;

exports[`super class method 1`] = `
"var v3 = {};
Object.setPrototypeOf(v3, Object.prototype);
var v2 = {};
Object.setPrototypeOf(v2, v3);
var v1 = { internal: \\"hello\\" };
Object.setPrototypeOf(v1, v2);
const v5 = (foo => () => foo.foo())(v1);
exports.handler = v5"
`;

exports[`value captured multiple times is only emitted once 1`] = `
"const v2 = (() => function bar() {
    return \\"hello\\";
})();
const v3 = (foo => () => foo)([v2, v2, v2]);
exports.handler = v3"
`;
